### PR TITLE
Run emscripten tests on ci

### PR DESCRIPTION
**Pull request by madebr**
- Add a `--no-threads` option to tests that use threads optionally, and disable thread tests that only test threads (on platforms that don't support them out-of-the-box)
- Add a `--map` argument to the `test/emscripten/server.py` mini file server. This is useful when combined with building emscripten with `-gsource-map -ffile-prefix-map=${{ github.workspace }}=/SDL --source-map-base /`.
By running the server with `--map "/SDL:/path/to/source/of/SDL3`, chome developer tools can find the sources.
- Run the emscripten tests driven by python3-selenium. A few tests currently fail, which I don't know how to fix.

The emscripten tests accept a few query arguments:
- `onassert` to set the default action when an assertion action (to avoid the browser showing a dialog)
- `loghtml` to do stdout to html element in addition to the console. This is used by selenium to get stdout.
- `arg_XXX` arguments to pass arguments

### CODE REVIEW

**Comment by ericoporto** [comment]
This is great work but I don't think there is anything wrong with your code in particular, afaict the failing tests are really failing.

**Comment by slouken** [comment]
Please don't exclude failing tests here unless they're tests we're comfortable having the platform fail (and document in the API functions), we want to make sure we don't forget about them.

**Comment by kg** [comment]
From a quick glance at the logs:
* TestVirtualJoystick looks like it's failing because it assumes joysticks will have names, but the web platform is aggressively anti-fingerprinting and doesn't expose joystick names. This is something you'd probably want to change the test to not assert on the web.
* testSetError is strange and looks like maybe a real bug in the browser version of SDL.
* addRemoveTimer looks like it probably assumes a native threading model, where you can set a timer, synchronously wait for it to expire, then assume the callback will have run. browser timers don't work that way.
* setWindowCenteredOnDisplay assumes that you can just open and position/size a "window" for a web game. this is technically true, but not usually what people would want. it makes sense to me that this test would fail in the browser and it should probably just be disabled. it's the wrong model for games that run inside a tab.

**Comment by madebr** [comment]
The `TestVirtualJoystick` test fails here:
https://github.com/libsdl-org/SDL/blob/0f27686a5125b08b59f2adb91820f12b53a62e04/test/testautomation_joystick.c#L67

It returns `"Standard Gamepad"` where it expects `"Virtual NVIDIA SHIELD Controller"`.
It think (but am not sure) `"Standard Gamepad"` comes from the gamepad database:
https://github.com/libsdl-org/SDL/blob/0f27686a5125b08b59f2adb91820f12b53a62e04/src/joystick/SDL_gamepad_db.h#L868

Is this expected behavior? @slouken

**Comment by ericoporto** [comment]
@madebr do you know if `SDL_JOYSTICK_VIRTUAL` is indeed defined for Emscripten?

**Comment by madebr** [comment]
> `SDL_JOYSTICK_VIRTUAL`

I think so locally, adding the following to a file fails the build:
```c
#ifdef SDL_JOYSTICK_VIRTUAL
blah
#endif
```

**Comment by slouken** [comment]
> The `TestVirtualJoystick` test fails here:
> 
> https://github.com/libsdl-org/SDL/blob/0f27686a5125b08b59f2adb91820f12b53a62e04/test/testautomation_joystick.c#L67
> 
> `SDL_GetGamepadName` returns `"Standard Gamepad"` where the test expects `"Virtual NVIDIA SHIELD Controller"`. It think (but am not sure) `"Standard Gamepad"` comes from the gamepad database:
> 
> https://github.com/libsdl-org/SDL/blob/0f27686a5125b08b59f2adb91820f12b53a62e04/src/joystick/SDL_gamepad_db.h#L867-L869
> 
> I added breakpoints in all emscripten functions in `src/joystick/emscripten/SDL_sysjoystick.c`, and only got hits in `EMSCRIPTEN_JoystickGetCount`, `EMSCRIPTEN_JoystickInit` and `EMSCRIPTEN_JoystickQuit`. I don't have a joystick connected.
> 
> Is this expected behavior? @slouken

I wouldn't have expected the default mapping to override the virtual gamepad mapping, but that's not a commonly tested code path and likely a legitimate bug. We should be able to reproduce that outside of emscripten by adding a default mapping for another platform. Can you create a separate issue for this so we can track it?

**Comment by ericoporto** [comment]
Could you please add it here too https://github.com/libsdl-org/SDL/blob/c244eb937b7fb31202580f2a28f2e81e706f861e/include/build_config/SDL_build_config_emscripten.h to keep in sync with the other build_configs ?

**Comment by slouken** [comment]
> Could you please add it here too https://github.com/libsdl-org/SDL/blob/c244eb937b7fb31202580f2a28f2e81e706f861e/include/build_config/SDL_build_config_emscripten.h to keep in sync with the other build_configs ?

Added in fa5367d37.

**Comment by slouken** [comment]
@madebr, are there any tests still failing or is this safe to merge?

**Comment by ericoporto** [comment]
Better just rebase it before merging.

**Comment by madebr** [comment]
> @madebr, are there any tests still failing or is this safe to merge?

This needs a rebase.
I'll have to retest to see what is still failing.

File: include/SDL3/SDL_main.h
Comment (madebr) [inline include/SDL3/SDL_main.h:98]: @icculus 
This undoes a change in https://github.com/libsdl-org/SDL/pull/10171#discussion_r1720765567, and is not a fix.

Doing `SDL_MAIN_HANDLED` in the tests (for emscripten only) would also work, but that is also not a fix.

**Comment by madebr** [comment]
testautomation segfaults somewhere in `SDL_DestroyCursor`.
I cannot reproduce the error on my system.

It's perhaps noteworthy the emscripten tests runs with the emscripten audio and video driver, whereas the other platforms run with the dummy drivers under ctest.

**Comment by madebr** [comment]
Can somebody take a look at the `SDL_DestroyCursor` issue?

I think the issue is a double-free, or free-ing static memory.

I pushed [this commit for some printf-debugging](https://github.com/madebr/SDL/commit/00e90000a661ec0abb37d5891d343d8f6773ff24#diff-41bdc236909c8708ff058aa571192a0db6e42e4373f12011b7dd3171bc2e586a), which produces [this output on ci](https://github.com/madebr/SDL/actions/runs/10436776974/job/28902244467):
```
08/18/24 01:41:50: Assert 'About to call SDL_DestroyCursor()': [0;32mPassed[0m
Start of SDL_DestroyCursor
SDL_DestroyCursor: for-loop iteration
SDL_DestroyCursor: About to call mouse->FreeCursor (in loop) 0xd7f5f0
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:154 Start of Emscripten_FreeCursor
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:156 Emscripten_FreeCursor: cursor is not null!
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:161 About to do SDL_free((char *)curdata->system_cursor)
/SDL/src/SDL.c SDL_QuitSubSystem:508 Start of SDL_QuitSubSystem
/SDL/src/SDL.c SDL_QuitSubSystem:508 Start of SDL_QuitSubSystem
/SDL/src/SDL.c SDL_QuitSubSystem:597 SDL_QuitSubSystem, Before SDL_QuitEvents()
/SDL/src/SDL.c SDL_QuitSubSystem:599 SDL_QuitSubSystem, After SDL_QuitEvents()
/SDL/src/SDL.c SDL_QuitSubSystem:603 End of SDL_QuitSubSystem
/SDL/src/SDL.c SDL_QuitSubSystem:574 SDL_QuitSubSystem, Before SDL_VideoQuit
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:154 Start of Emscripten_FreeCursor
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:156 Emscripten_FreeCursor: cursor is not null!
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:164 About to do SDL_free(cursor->internal)
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:168 About to do SDL_free(cursor)
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:171 End of Emscripten_FreeCursor
/SDL/src/SDL.c SDL_QuitSubSystem:508 Start of SDL_QuitSubSystem
/SDL/src/SDL.c SDL_QuitSubSystem:603 End of SDL_QuitSubSystem
/SDL/src/video/SDL_video.c SDL_VideoQuit:4247 Before SDL_EnableScreenSaver
/SDL/src/video/SDL_video.c SDL_VideoQuit:4254 Before VideoQuit
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_QuitMouse:255 Start of Emscripten_QuitMouse
/SDL/src/video/SDL_video.c SDL_VideoQuit:4256 After VideoQuit
/SDL/src/video/SDL_video.c SDL_VideoQuit:4260 SDL_DelVideoDisplay Iteration
/SDL/src/video/SDL_video.c SDL_VideoQuit:4265 Before SDL_free(_this->displays)
/SDL/src/video/SDL_video.c SDL_VideoQuit:4270 Before SDL_free(_this->primary_selection_text)
/SDL/src/video/SDL_video.c SDL_VideoQuit:4274 Before _this->free(_this)
/SDL/src/video/SDL_video.c SDL_VideoQuit:4276 End of SDL_VideoQuit
/SDL/src/SDL.c SDL_QuitSubSystem:576 SDL_QuitSubSystem, After SDL_VideoQuit
/SDL/src/SDL.c SDL_QuitSubSystem:578 SDL_QuitSubSystem, Before SDL_QuitSubSystem(SDL_INIT_EVENTS)
/SDL/src/SDL.c SDL_QuitSubSystem:508 Start of SDL_QuitSubSystem
/SDL/src/SDL.c SDL_QuitSubSystem:603 End of SDL_QuitSubSystem
/SDL/src/SDL.c SDL_QuitSubSystem:580 SDL_QuitSubSystem, After SDL_QuitSubSystem(SDL_INIT_EVENTS)
/SDL/src/SDL.c SDL_QuitSubSystem:588 SDL_QuitSubSystem, Before SDL_QuitTimers()
/SDL/src/SDL.c SDL_QuitSubSystem:590 SDL_QuitSubSystem, After SDL_QuitTimers()
/SDL/src/SDL.c SDL_QuitSubSystem:603 End of SDL_QuitSubSystem
INFO:__main__:Exit code = 1
```
The following sequence is weird.
```
/SDL/src/video/emscripten/SDL_emscriptenmouse.c Emscripten_FreeCursor:161 About to do SDL_free((char *)curdata->system_cursor)
/SDL/src/SDL.c SDL_QuitSubSystem:508 Start of SDL_QuitSubSystem
```
https://github.com/madebr/SDL/blob/00e90000a661ec0abb37d5891d343d8f6773ff24/src/video/emscripten/SDL_emscriptenmouse.c#L161-L164
There should be `About to do SDL_free(cursor->internal)` instead.

I can reproduce this by adding this code to the test case:
```c
void *p;
p = SDL_malloc(128);
SDL_free(p);
SDL_free(p);
```
or
```c
 SDL_free((void*)(uintptr_t)0x46789);
```

**Comment by icculus** [comment]
> Can somebody take a look at the SDL_DestroyCursor issue?

Did we fix this already? Running testautomation.html from this PR's branch on both Firefox and Chrome (linux/amd64), I don't get a crash.

```
Run Summary: Total=342 Passed=330 Failed=0 Skipped=12
```

**Comment by ericoporto** [comment]
@icculus its still erroring, the problem is it doesn't reproduce in Chrome or Firefox locally, check the CI log

```
'mouse_createFreeCursor' started[0m
2024-08-19T18:41:48.6654235Z 2: 08/19/24 18:41:44: Test Description: 'Check call to SDL_CreateCursor and SDL_DestroyCursor'
2024-08-19T18:41:48.6654606Z 2: 08/19/24 18:41:44: Test Iteration 1: execKey 17353597579503114674
2024-08-19T18:41:48.6655157Z 2: 08/19/24 18:41:44: Assert 'Call to SDL_CreateCursor()': [0;32mPassed[0m
2024-08-19T18:41:48.6656133Z 2: 08/19/24 18:41:44: Assert 'Validate result from SDL_CreateCursor() is not NULL': [0;32mPassed[0m
2024-08-19T18:41:48.6656763Z 2: 08/19/24 18:41:44: Assert 'About to call SDL_DestroyCursor()': [0;32mPassed[0m
2024-08-19T18:41:48.6656940Z 2: INFO:__main__:Exit code = 1
2024-08-19T18:41:48.6657358Z 11/20 Test  #2: testautomation ...................***Failed   17.82 sec
```

**Comment by madebr** [comment]
> > Can somebody take a look at the SDL_DestroyCursor issue?
> 
> Did we fix this already? Running testautomation.html from this PR's branch on both Firefox and Chrome (linux/amd64), I don't get a crash.
> 
> ```
> Run Summary: Total=342 Passed=330 Failed=0 Skipped=12
> ```

I only happens on ci.
No idea how to reproduce (https://github.com/libsdl-org/SDL/pull/9937#issuecomment-2295064383 does not work anymore)
Also, a double-free is less likely because ci is run with `--trackmem`.
This means SDL_test should detect already-freed memory.


The differences between my system and ci are:
- Ubuntu 22.04
- newest chrome
- testing on a 100% headless system

Using latest emsdk does not change the failure mode.

**Comment by icculus** [comment]
Emscripten_FreeCursor is calling `SDL_free((char *)curdata->system_cursor);` but if this came from Emscripten_CreateCursor(), it allocated that buffer with malloc(), not SDL_malloc().

Likely culprit...?

**Comment by madebr** [comment]
> Likely culprit...?

Yup, [changing `SDL_free((char *)curdata->` into `free((char *)curdata->`](https://github.com/madebr/SDL/commit/01d926046e655a444ebf61a3ab393a3cbd3de0d9) fixes the weird behavior (undefined behavior ftw).
https://github.com/madebr/SDL/actions/runs/10460083869/job/28965588919#step:24:12273

This is not a proper patch. Is there a way to call SDL_free from within the embedded javascript code?

**Comment by icculus** [comment]
See if 423d6ec15a9efa613b0e611e4708d7df95214289 fixes it.

**Comment by madebr** [comment]
> See if [423d6ec](https://github.com/libsdl-org/SDL/commit/423d6ec15a9efa613b0e611e4708d7df95214289) fixes it.

It does!
Thanks for investing your time into this!

I've removed the `SDL_ASSERT` workaround, and marked this pr ready for review.

**Review by slouken — APPROVED** [review]

**Comment by slouken** [comment]
Looks good! Thanks for all your work here!

**Comment by ericoporto** [comment]
@madebr amazing work! Incredible attention to detail. This PR looks great!

**Comment by madebr** [comment]
> @madebr amazing work! Incredible attention to detail. This PR looks great!

Thanks, but I couldn't do it without the help of Ryan, and many others.


I applied a little code reformatting, and will merge it after ci has its say.

**Pull request merged**
